Initialise release script

juliaogris opened the pull request:
Fixes #45.

Changes proposed in this pull request:
- Create release script (`src/scripts/release.sh <prepare|deploy> X.Y.Z`)
- Release preparation via release.sh: `release.sh prepare X.Y.Z`
  * Create release branch
  * Update, commit and push version file
  * Create PR from `origin/release-<VERSION>` to `upstream/master`
- Release deployment via release.sh: `release.sh deploy X.Y.Z`
  * Tag current version
  * Push tags to `updastream/master`
- Travis and Appveyor config update to deploy to `Github releases` on tags

@anz-bank/sysl-developers

camh-anz commented:
There is a linter called shellcheck that is available in homebrew.

Please run this and fix any warnings/errors it emits.

Shellcheck can be conservative sometimes because it's hard to tell if you want to do something that is normally avoided. You can silence the warnings by putting in a comment:

`# shellcheck disable=SC2032`

where SC2032 is the code for the warning emitted.

Let me know when you've done this and I'll review it.

anzdaddy commented on src/scripts/release.sh:
```diff
@@ -0,0 +1,74 @@
+    echo "$GITHUB_PR_URL"
+    open "$GITHUB_PR_URL"
+elif [ $WGET_STATUS -eq 6 ]; then
+	exit_error "Wrong username or password/token."
```
ditto

juliaogris replied:
Done.

juliaogris commented on src/scripts/release.sh:
```diff
@@ -0,0 +1,74 @@
+COMMAND=$1
+VERSION=$2
+
+function exit_error {
```
Done

juliaogris commented on src/scripts/release.sh:
```diff
@@ -0,0 +1,74 @@
+git co -b $RELEASE_BRANCH || exit_error "Cannot create release branch $RELEASE_BRANCH"
+
+echo "------- Update version ---------"
+echo "__version__ = '$VERSION'" > 'src/sysl/__version__.py' || { echo "Could not override src/sysl/__version__.py"; exit 1; }
```
I used single command `fatal` instead.

juliaogris commented on src/scripts/release.sh:
```diff
@@ -0,0 +1,74 @@
+
+echo "------- Create PR ---------"
+REPO="sysl"
+CREDENTIALS=`echo "host=github.com" | git credential fill`
```
Done.

juliaogris commented on src/scripts/release.sh:
```diff
@@ -0,0 +1,74 @@
+    exit_error "Could not get GitHub credentials, please create pull request manually."
+fi
+
+JSON=`jq -nc "{title:\"Bump version to $VERSION\",head:\"$ORIGIN:$RELEASE_BRANCH\",base:\"master\"}"`
```
Done (removed).

juliaogris commented on src/scripts/release.sh:
```diff
@@ -0,0 +1,74 @@
+
+WGET_STATUS=$?
+if [ $WGET_STATUS -eq 0 ]; then
+	GITHUB_PR_URL=`echo "$RESPONSE" | jq -r '.html_url'`
```
Done.

juliaogris commented on src/scripts/release.sh:
```diff
@@ -0,0 +1,74 @@
+CREDENTIALS=`echo "host=github.com" | git credential fill`
+USERNAME=`echo $CREDENTIALS | sed -n "s/.*username=\([^ ]*\).*/\1/p"`
+PASSWORD=`echo $CREDENTIALS | sed -n "s/.*password=\([^ ]*\).*/\1/p"`
+if [ -z "$USERNAME" ] || [ -z "$PASSWORD" ]; then
```
Apologies I'm reverting this. `shellcheck` says  `SC2166: Prefer [ p ] || [ q ] as [ p -o q ] is not well defined.`

anzdaddy commented on src/scripts/release.sh:
```diff
@@ -0,0 +1,74 @@
+
+WGET_STATUS=$?
+if [ $WGET_STATUS -eq 0 ]; then
+	  GITHUB_PR_URL=$(echo "$RESPONSE" | jq -r '.html_url')
```
Indent still broken. You're mixing tabs and spaces, so it might not be evident in your editor.

juliaogris replied:
Ok, sorry, thought the linter would pick this up.
All spaces now.

anzdaddy commented on src/scripts/release.sh:
```diff
@@ -0,0 +1,74 @@
+if [[ $COMMAND != "prepare" && $COMMAND != "deploy" ]]; then
+	fatal "$USAGE"
+fi
+if [[ $(git status --porcelain 2> /dev/null | tail -n1) != "" ]]; then
```
Sorry, didn't pick this up earlier. Either use `-n` here or `= ""` instead of `-z` below.

juliaogris replied:
Done (changed the bottom one).

juliaogris commented:
@camh-anz spellcheck reports no warning with `spellcheck realse.sh` (thanks for that tip - love it!)
@anzdaddy - I believe I've addressed all comments, please re-review.
Thank you both!

juliaogris commented:
@anzdaddy, sorry indentation is still bad - on sec!

juliaogris commented:
Alright @anzdaddy, good to go (hopefully 🙀).

juliaogris commented:
@camh-anz, @anzdaddy - bad etiquette on my behalf, I've just made this PR a bit bigger by adding the tagging/deployment to the release script and updating the CI configs (Tavis & Appveyor). 
Please re-review.

juliaogris commented on .travis.yml:
```diff
@@ -1,30 +1,46 @@
 language: python
```
`travis setup releases` as suggested on the [Travis docs](https://docs.travis-ci.com/user/deployment/releases/)
is responsible for the whitespace changes.

juliaogris commented on src/sysl/__version__.py:
```diff
@@ -1,3 +1 @@
-VERSION = (0, 1, 1)
-
-__version__ = '.'.join(map(str, VERSION))
+__version__ = '0.1.2'
```
Updated this for testing purposes.
